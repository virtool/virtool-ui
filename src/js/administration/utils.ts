import { Account } from "@account/types";
import { Permission } from "../groups/types";
import { AdministratorRoles } from "./types";

/**
 * The permissions level of each administrator role
 */
enum AdministratorPermissionsLevel {
    full,
    settings,
    spaces,
    users,
    base,
}

/**
 * Check if a user has a sufficient admin role
 *
 * @param requiredRole - The lowest admin role the user must have to pass the check
 * @param userRole - The administrator role of the user
 */

export function hasSufficientAdminRole(requiredRole: AdministratorRoles, userRole: AdministratorRoles): boolean {
    return AdministratorPermissionsLevel[userRole] <= AdministratorPermissionsLevel[requiredRole];
}

/**
 * Permissions granted to each administrator role
 */
export enum AdministratorPermissions {
    cancel_job = AdministratorRoles.BASE,
    create_ref = AdministratorRoles.BASE,
    modify_hmm = AdministratorRoles.BASE,
    remove_job = AdministratorRoles.BASE,
    upload_file = AdministratorRoles.FULL,
    create_sample = AdministratorRoles.FULL,
    modify_subtraction = AdministratorRoles.FULL,
    remove_file = AdministratorRoles.FULL,
}

/**
 * Check if a user has a sufficient admin role or legacy permissions to perform an action
 *
 * @param account - The Account object of the user
 * @param permission - The permissions to check
 * @returns  Whether the user is allowed to perform the action
 */
export function checkAdminRoleOrPermissionsFromAccount(account: Account, permission: Permission): boolean {
    return (
        hasSufficientAdminRole(AdministratorPermissions[permission as string], account.administrator_role) ||
        account.permissions[permission]
    );
}
