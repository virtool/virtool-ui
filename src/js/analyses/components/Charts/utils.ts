import { filter, reduce, zipWith } from "lodash-es";
import { IimiSequence, UntrustworthyRange } from "../../types";
/**
 * Convert range length encoded data into an array of depth values.
 *
 * @param lengths - the lengths of a given read depth region
 * @param rle - the depth values of the regions
 *
 * @returns An array of depth values
 */
export function convertRleToCoverage(lengths: Array<number>, rle: Array<number>) {
    const coverage = [];

    for (let sharedIndex = 0; sharedIndex < lengths.length; sharedIndex++) {
        const length = lengths[sharedIndex];
        const value = rle[sharedIndex];

        coverage.push(...Array(length).fill(value));
    }

    return coverage;
}

/**
 * Average the read depths of an array of sequences.
 *
 * @param sequences - the sequences to average
 * @returns An averaged sequence
 */
export function averageSequences(sequences: IimiSequence[]) {
    return zipWith(...sequences, (...args) => {
        const filtered = filter(args, num => typeof num === "number");
        return filtered.length === 0 ? 0 : reduce(filtered, (a, b) => a + b) / filtered.length;
    });
}

/**
 * Determine the regions of the sequence that are trustworthy using the untrustworthy regions.
 *
 * @param length - The length of the sequence
 * @param untrustworthyRanges - the untrustworthy regions to infer trustworthy regions from
 */
export function deriveTrustworthyRegions(length: number, untrustworthyRanges: [number, number][]): [number, number][] {
    const trustworthyRanges = [];
    let start = 1;
    let end;

    untrustworthyRanges.forEach(range => {
        end = range[0];
        trustworthyRanges.push([start, end]);
        start = range[1];
    });

    trustworthyRanges.push([start, length]);
    return trustworthyRanges;
}

/**
 * Construct a union of untrustworthy regions from multiple sources.
 *
 * @param untrustworthyRanges - A collection of untrustworthy regions to combine
 * @returns the combined untrustworthy regions
 */
export function combineUntrustworthyRegions(untrustworthyRanges: UntrustworthyRange[][]): UntrustworthyRange[] {
    const sortedUntrustworthyRanges = untrustworthyRanges.flat().sort((a, b) => a[0] - b[0]);

    if (sortedUntrustworthyRanges.length === 0) {
        return [];
    }

    const combined = [] as [number, number][];
    combined.push(sortedUntrustworthyRanges.shift());

    sortedUntrustworthyRanges.forEach(range => {
        const last = combined[combined.length - 1];

        if (range[0] <= last[1]) {
            last[1] = range[1];
        } else {
            combined.push(range);
        }
    });

    return combined;
}
