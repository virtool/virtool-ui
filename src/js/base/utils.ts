import { getColor, theme } from "@app/theme";

type getBadgeOrLabelColorProps = {
    color?: string;
};

export function getBadgeOrLabelColor(props: getBadgeOrLabelColorProps): string {
    if (props.color) {
        return getColor({ color: props.color, theme });
    }

    return theme.color.greyDark;
}
