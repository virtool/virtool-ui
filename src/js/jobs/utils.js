export const getStepDescription = ({ state, step_name, step_description }) => {
    if (step_name && step_description) {
        return {
            title: step_name,
            description: step_description
        };
    }

    if (state === "waiting") {
        return {
            title: "Waiting",
            description: "Waiting for resources to become available."
        };
    }

    if (state === "preparing") {
        return {
            description: "Preparing the workflow environment."
        };
    }

    if (state === "complete") {
        return {
            title: "Complete",
            description: ""
        };
    }

    if (state === "cancelled") {
        return {
            title: "Cancelled",
            description: ""
        };
    }

    if (state === "terminated") {
        return {
            title: "Terminated",
            description: "There was a system malfunction"
        };
    }
    return {
        title: "",
        description: ""
    };
};
