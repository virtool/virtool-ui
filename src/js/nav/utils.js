import { includes } from "lodash-es";

export function excludePaths(paths = []) {
    return function (match, location) {
        if (includes(paths, location.pathname)) {
            return false;
        }

        return Boolean(match);
    };
}
