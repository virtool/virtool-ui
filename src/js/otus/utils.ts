import { OTUSegment, OTUSequence } from "@otus/types";
import { map } from "lodash";
import { indexOf, sortBy } from "lodash-es";

/**
 * A hook for sorting the sequences for the active isolate
 *
 * @param sequences - The active isolate sequences
 * @param segments - The segments associated with the OTU
 */
export default function sortSequencesBySegment(
    sequences: OTUSequence[],
    segments: OTUSegment,
): OTUSequence[] {
    if (sequences) {
        const segmentNames = map(segments, "name");
        return sortBy(sequences, [
            (entry) => {
                const index = indexOf(segmentNames, entry.segment);
                return index !== -1 ? index : segmentNames.length;
            },
        ]);
    }

    return [];
}
