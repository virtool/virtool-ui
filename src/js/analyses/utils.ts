import {
    compact,
    fill,
    filter,
    flatMap,
    fromPairs,
    has,
    includes,
    keys,
    map,
    max,
    maxBy,
    min,
    minBy,
    reject,
    sortBy,
    startsWith,
    sumBy,
    toNumber,
    uniq,
    zipWith,
} from "lodash-es";
import { cloneDeep } from "lodash-es/lodash";
import { formatIsolateName } from "../utils/utils";
import { PositionMappedReadDepths, UntrustworthyRange } from "./types";

export const calculateAnnotatedOrfCount = orfs => filter(orfs, orf => orf.hits.length).length;

function calculateORFMinimumE(hits) {
    if (hits.length === 0) {
        return;
    }

    const minHit = minBy(hits, "full_e");
    return minHit.full_e;
}

function calculateSequenceMinimumE(orfs) {
    if (orfs.length === 0) {
        return;
    }

    const minEValues = map(orfs, orf => calculateORFMinimumE(orf.hits));
    return min(minEValues);
}

export function extractFamilies(orfs) {
    const families = uniq(flatMap(orfs, orf => flatMap(orf.hits, hit => keys(hit.families))));
    return reject(families, f => f === "None");
}

export function extractNames(orfs) {
    return uniq(flatMap(orfs, orf => flatMap(orf.hits, hit => hit.names)));
}

/**
 * Transform an array of coordinate pairs into an flat array where the index is the x coordinate and the value is the y
 * coordinate.
 *
 * @param {Array} align - the coordinates
 * @param {Number} length - the length of the generated flat array
 * @returns {Array} - the flat array
 *
 */
export const fillAlign = ({ align, length }) => {
    const filled = Array(length);

    if (!align) {
        return fill(filled, 0);
    }

    const coords = fromPairs(align);

    let prev = 0;

    return map(filled, (depth, i) => {
        if (has(coords, i)) {
            prev = coords[i];
        }

        return prev;
    });
};

function getIdentities(data) {
    return flatMap(data, item => item.identities);
}

const getSequenceIdentities = sequence => flatMap(sequence.hits, hit => hit.identity);

export const formatAODPData = detail => {
    const results = map(detail.results, result => {
        const isolates = map(result.isolates, isolate => {
            const sequences = map(isolate.sequences, sequence => {
                return {
                    ...sequence,
                    identities: getSequenceIdentities(sequence),
                };
            });

            return {
                ...isolate,
                sequences,
                identities: getIdentities(sequences),
            };
        });

        const identities = getIdentities(isolates);

        return {
            ...result,
            isolates,
            identities,
            identity: max(identities),
        };
    });

    return { ...detail, results };
};

export const formatNuVsData = detail => {
    const hits = map(detail.results.hits, hit => ({
        ...hit,
        id: toNumber(hit.index),
        annotatedOrfCount: calculateAnnotatedOrfCount(hit.orfs),
        e: calculateSequenceMinimumE(hit.orfs),
        families: extractFamilies(hit.orfs),
        names: extractNames(hit.orfs),
    }));

    const longestSequence = maxBy(hits, hit => hit.sequence.length);

    const { cache, created_at, id, ready, user, workflow } = detail;

    return {
        cache,
        created_at,
        id,
        ready,
        results: {
            ...detail.results,
            hits,
        },
        user,
        workflow,
        maxSequenceLength: longestSequence.sequence.length,
    };
};

/**
 * Calculate the median of an Array of numbers.
 *
 * @param values - an array of numbers
 * @returns {number|*} - the median
 */
export const median = values => {
    const sorted = values.slice().sort((a, b) => a - b);

    const midIndex = (sorted.length - 1) / 2;

    if (midIndex % 1 === 0) {
        return sorted[midIndex];
    }

    const lowerIndex = Math.floor(midIndex);
    const upperIndex = Math.ceil(midIndex);

    return Math.round((sorted[lowerIndex] + sorted[upperIndex]) / 2);
};

/**
 * Merge the coverage arrays for the given isolates. This is used to render a representative coverage chart for the
 * parent OTU.
 *
 * @param isolates
 * @returns {Array}
 */
export const mergeCoverage = isolates => {
    const longest = maxBy(isolates, isolate => isolate.filled.length);
    const coverages = map(isolates, isolate => isolate.filled);
    return map(longest.filled, (depth, index) => max(map(coverages, coverage => coverage[index])));
};

export const formatSequence = (sequence, readCount) => ({
    ...sequence,
    filled: fillAlign(sequence),
    reads: sequence.pi * readCount,
});

export const formatPathoscopeData = detail => {
    if (detail.results.hits.length === 0) {
        return detail;
    }

    const { cache, created_at, results, id, index, ready, reference, subtractions, user, workflow } = detail;

    const readCount = results.read_count;

    const hits = map(results.hits, otu => {
        const isolateNames = [];

        // Go through each isolate associated with the OTU, adding properties for weight, read count,
        // median depth, and coverage. These values will be calculated from the sequences owned by each isolate.
        const isolates = map(otu.isolates, isolate => {
            // Make a name for the isolate by joining the source type and name, eg. "Isolate" + "Q47".
            const name = formatIsolateName(isolate);

            isolateNames.push(name);

            const sequences = sortBy(
                map(isolate.sequences, sequence => formatSequence(sequence, readCount)),
                "length"
            );

            const filled = flatMap(sequences, "filled");

            // Coverage is the number of non-zero depth positions divided by the total number of positions.
            const coverage = compact(filled).length / filled.length;

            return {
                ...isolate,
                name,
                filled,
                coverage,
                sequences,
                maxDepth: max(filled),
                pi: sumBy(sequences, "pi"),
                depth: median(filled),
            };
        });

        const filled = mergeCoverage(isolates);
        const pi = sumBy(isolates, "pi");

        return {
            ...otu,
            filled,
            pi,
            isolates: sortBy(isolates, "coverage").reverse(),
            coverage: maxBy(isolates, "coverage").coverage,
            depth: median(filled),
            isolateNames: reject(uniq(isolateNames), "Unnamed Isolate"),
            maxGenomeLength: maxBy(isolates, "filled.length").length,
            maxDepth: maxBy(isolates, "maxDepth").maxDepth,
            reads: pi * readCount,
        };
    });

    return {
        cache,
        created_at,
        id,
        index,
        reference,
        ready,
        results: {
            hits,
            readCount,
            subtractedCount: detail.results.subtracted_count,
        },
        subtractions,
        user,
        workflow,
    };
};

export const fuseSearchKeys = {
    pathoscope_bowtie: ["name", "abbreviation"],
    nuvs: ["families", "names"],
    aodp: ["name"],
};

export const formatData = detail => {
    if (startsWith(detail?.workflow, "pathoscope")) {
        return formatPathoscopeData(detail);
    }

    if (detail?.workflow === "nuvs") {
        return formatNuVsData(detail);
    }

    if (detail?.workflow === "aodp") {
        return formatAODPData(detail);
    }

    return detail;
};

const supportedWorkflows = ["pathoscope_bowtie", "nuvs", "aodp", "iimi"];

export function checkSupportedWorkflow(workflow) {
    return includes(supportedWorkflows, workflow);
}

/**
 * Convert range length encoded data into an array of depth values.
 *
 * @param lengths - the lengths of a given read depth region
 * @param rle - the depth values of the regions
 *
 * @returns An array of depth values
 */
export function convertRleToCoverage(lengths: Array<number>, rle: Array<number>) {
    const coverage = [];

    for (let sharedIndex = 0; sharedIndex < lengths.length; sharedIndex++) {
        const length = lengths[sharedIndex];
        const value = rle[sharedIndex];

        coverage.push(...Array(length).fill(value));
    }

    return coverage;
}

/**
 * Average the read depths of an array of sequences.
 *
 * @param sequences - the sequences to average
 * @returns An averaged sequence
 */
export function maxSequences(sequences: PositionMappedReadDepths[]) {
    return zipWith(...sequences, (...args) => max(args));
}

/**
 * Determine the regions of the sequence that are trustworthy using the untrustworthy regions.
 *
 * @param length - The length of the sequence
 * @param untrustworthyRanges - the untrustworthy regions to infer trustworthy regions from
 */
export function deriveTrustworthyRegions(length: number, untrustworthyRanges: [number, number][]): [number, number][] {
    const trustworthyRanges = [];
    let start = 1;
    let end;

    untrustworthyRanges.forEach(range => {
        end = range[0];
        trustworthyRanges.push([start, end]);
        start = range[1];
    });

    trustworthyRanges.push([start, length]);
    return trustworthyRanges;
}

/**
 * Construct a union of untrustworthy regions from multiple sources.
 *
 * @param untrustworthyRanges - A collection of untrustworthy regions to combine
 * @returns the combined untrustworthy regions
 */
export function combineUntrustworthyRegions(untrustworthyRanges: UntrustworthyRange[][]): UntrustworthyRange[] {
    untrustworthyRanges = cloneDeep(untrustworthyRanges);

    const sortedUntrustworthyRanges = untrustworthyRanges.flat().sort((a, b) => a[0] - b[0]);

    if (sortedUntrustworthyRanges.length === 0) {
        return [];
    }

    const combined = [sortedUntrustworthyRanges.shift()] as [number, number][];

    sortedUntrustworthyRanges.forEach(range => {
        const last = combined[combined.length - 1];

        if (range[0] <= last[1]) {
            last[1] = range[1];
        } else {
            combined.push(range);
        }
    });

    return combined;
}
