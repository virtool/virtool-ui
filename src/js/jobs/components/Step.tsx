import { format } from "date-fns";
import React from "react";
import styled from "styled-components";
import { BoxGroupSection, Icon, Loader } from "../../base";
import { JobState } from "../types";
import { getStepDescription } from "../utils";
import { JobStateIcon } from "./JobStateIcon";

const JobStepLoader = styled(Loader)`
    padding: 0 1.5px;
`;

const StyledJobStepTimestamp = styled.div`
    align-items: center;
    display: flex;
    margin-top: 10px;

    i:not(:first-child) {
        padding-left: 10px;
    }

    & > span {
        padding-left: 5px;
    }
`;

interface JobStepTimestampProps {
    timestamp: string;
}

function JobStepTimestamp({ timestamp }: JobStepTimestampProps) {
    return (
        <StyledJobStepTimestamp>
            <Icon name="clock" />
            <span>{format(new Date(timestamp), "HH:mm:ss")}</span>
            <Icon name="calendar" />
            <span>{format(new Date(timestamp), "yyyy-MM-dd")}</span>
        </StyledJobStepTimestamp>
    );
}

const StyledJobStepDescription = styled.div`
    display: flex;
    flex-direction: column;

    > h4 {
        font-size: ${props => props.theme.fontSize.lg};
        font-weight: bold;
        margin: 0 0 4px;
    }

    > p {
        color: ${props => props.theme.color.greyDarkest};
        margin: 0 0 3px;
    }
`;

type JobStep = {
    state: JobState;
    step_description: string;
    step_name: string;
    timestamp: string;
};

interface JobStepDescriptionProps {
    step: JobStep;
}

function JobStepDescription({ step }: JobStepDescriptionProps) {
    const { description, title } = getStepDescription(step);

    return (
        <StyledJobStepDescription>
            <h4>{title}</h4>
            <p>{description}</p>
            <JobStepTimestamp timestamp={step.timestamp} />
        </StyledJobStepDescription>
    );
}

interface JobStepIconProps {
    complete: boolean;
    state: JobState;
}

function JobStepIcon({ complete, state }: JobStepIconProps) {
    if (complete) {
        return <Icon name="arrow-circle-down" color="blue" title={state} fixedWidth />;
    }

    if (state === "running") {
        return <JobStepLoader size="12px" color="blue" />;
    }

    return <JobStateIcon state={state} />;
}

const JobStepIconContainer = styled.div`
    align-items: center;
    display: flex;
    height: 16px;
    margin-right: 4px;
    padding-top: 3px;
    width: 16px;
`;

const StyledJobStep = styled(BoxGroupSection)`
    align-items: flex-start;
    display: flex;
`;

interface JobStepProps {
    complete: boolean;
    step: JobStep;
}

export function JobStep({ complete, step }: JobStepProps) {
    return (
        <StyledJobStep>
            <JobStepIconContainer>
                <JobStepIcon state={step.state} complete={complete} />
            </JobStepIconContainer>
            <JobStepDescription step={step} />
        </StyledJobStep>
    );
}

StyledJobStep.displayName = "JobStep";
