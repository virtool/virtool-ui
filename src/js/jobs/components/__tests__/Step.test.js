jest.mock("../../utils");
import { screen } from "@testing-library/react";
import * as functions from "../../utils";
import { JobStepDescription, JobStepIcon } from "../Step";

describe("<JobStepDescription />", () => {
    let props;
    const spyStepDescription = jest.spyOn(functions, "getStepDescription");
    beforeEach(() => {
        spyStepDescription.mockClear().mockReturnValue({ description: "foo", title: "bar" });
        props = {
            step_description: "foo",
            step_name: "bar",
            title: "test",
            timestamp: "2022-05-19T17:48:25.508000Z"
        };
    });

    it("renders correctly", () => {
        renderWithProviders(<JobStepDescription {...props} />);
        expect(screen.getByText("foo")).toBeInTheDocument();
        expect(screen.getByText("bar")).toBeInTheDocument();
    });

    it("getStepDescription is called 1 time", () => {
        renderWithProviders(<JobStepDescription {...props} />);
        expect(functions.getStepDescription).toHaveBeenCalledTimes(1);
    });

    it("getStepDescription returns correct values", () => {
        let step = props;
        let expectedVal = { description: "foo", title: "bar" };
        expect(functions.getStepDescription(step)).toEqual(expectedVal);
    });
});

describe("<JobStepIcon />", () => {
    let props;

    beforeEach(() => {
        props = {
            complete: false,
            state: "running"
        };
    });

    it("renders correct icon when step complete", () => {
        props.complete = true;
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state complete", () => {
        props.state = "complete";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state preparing", () => {
        props.state = "preparing";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state running", () => {
        props.state = "running";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state cancelled", () => {
        props.state = "cancelled";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state error", () => {
        props.state = "error";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders correct icon when state waiting", () => {
        props.state = "waiting";
        const wrapper = shallow(<JobStepIcon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
