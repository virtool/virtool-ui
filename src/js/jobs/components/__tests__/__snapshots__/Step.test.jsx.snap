// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<JobStep /> > should render text and icon when state is special case 1`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP iFSptS  fas fa-hourglass-end"
        color="red"
        title="timeout"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Timed Out
      </h4>
      <p>
        The job timed out.
      </p>
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 2`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP iFSptS  fas fa-skull"
        color="red"
        title="terminated"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Terminated
      </h4>
      <p>
        There was a system malfunction
      </p>
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 3`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP eAOQkA  fas fa-check"
        color="green"
        title="complete"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Complete
      </h4>
      <p />
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 4`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP iFSptS  fas fa-exclamation-triangle"
        color="red"
        title="error"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Error
      </h4>
      <p />
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 5`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP lbuzuw  fas fa-boxes"
        color="grey"
        title="preparing"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Preparing
      </h4>
      <p>
        Preparing the workflow environment.
      </p>
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 6`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP jjOKAv  fas fa-clock"
        color="purple"
        title="waiting"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Waiting
      </h4>
      <p>
        Waiting for resources to become available.
      </p>
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<JobStep /> > should render text and icon when state is special case 7`] = `
<DocumentFragment>
  <div
    class="sc-eDWCr sc-eYqcxL gmERkp fKKldT"
  >
    <div
      class="sc-fmPOXC byCizZ"
    >
      <i
        class="sc-hLBbgP iFSptS  fas fa-hourglass-end"
        color="red"
        title="timeout"
      />
    </div>
    <div
      class="sc-gkSfol cphaBg"
    >
      <h4>
        Timed Out
      </h4>
      <p>
        The job timed out.
      </p>
      <div
        class="sc-gzrROc gzboQX"
      >
        <i
          class="sc-hLBbgP icWVMU  fas fa-clock"
        />
        <span>
          17:48:05
        </span>
        <i
          class="sc-hLBbgP icWVMU  fas fa-calendar"
        />
        <span>
          2022-05-19
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
